### ISSUE TITLE
Random occurrences of "not a tree object" on fetch

### ISSUE THREAD

**Issue by dalemyers** [git-lfs/git-lfs#2974]
As part of our builds running on VSTS, the first thing it does is to clone our repo and checkout the commit we are building. Sometimes though, a few times in a hundred, we get a failure like this:

```
2018-04-04T09:05:06.4938770Z ##[command]git -c http.extraheader="AUTHORIZATION: bearer ********" fetch --tags --prune --progress --no-recurse-submodules origin
2018-04-04T09:05:06.9741870Z From https://[redacted].visualstudio.com/[redacted]/_git/[redacted]
2018-04-04T09:05:06.9756220Z  - [deleted]               (none)     -> origin/[redacted]
2018-04-04T09:05:07.2478720Z remote: Microsoft (R) Visual Studio (R) Team Services        
2018-04-04T09:05:07.3977900Z remote: 
2018-04-04T09:05:07.3994020Z remote: Found 14 objects to send. (150 ms)        
2018-04-04T09:05:07.6351740Z    e7989e48b8..8ec149e468  [redacted] -> origin/[redacted]
2018-04-04T09:05:07.6367720Z    9f00bb1ce0..caf9ab3dd0  develop             -> origin/develop
2018-04-04T09:05:07.7022600Z ##[command]git -c http.https://[redacted].visualstudio.com.extraheader="AUTHORIZATION: bearer ********" lfs fetch origin e0b225c0e5c2747b2ab4270238ef783113fcd673
2018-04-04T09:05:07.7452330Z fetch: Fetching reference e0b225c0e5c2747b2ab4270238ef783113fcd673
2018-04-04T09:05:07.7549930Z Could not scan for Git LFS files
2018-04-04T09:05:07.7618340Z 
2018-04-04T09:05:07.7635280Z Errors logged to /Users/[redacted]/vsts_agents/vsts_agent_2/_work/9/s/.git/lfs/logs/20180404T090507.754105495.log
2018-04-04T09:05:07.7649040Z Use `git lfs logs last` to view the log.
2018-04-04T09:05:07.7736300Z ##[command]git lfs logs last
2018-04-04T09:05:07.8064290Z git-lfs/2.4.0 (GitHub; darwin amd64; go 1.10)
2018-04-04T09:05:07.8078950Z git version 2.16.2
2018-04-04T09:05:07.8085600Z 
2018-04-04T09:05:07.8098800Z $ git-lfs fetch origin e0b225c0e5c2747b2ab4270238ef783113fcd673
2018-04-04T09:05:07.8112210Z fetch: Fetching reference e0b225c0e5c2747b2ab4270238ef783113fcd673
2018-04-04T09:05:07.8126000Z Could not scan for Git LFS files
2018-04-04T09:05:07.8132650Z 
2018-04-04T09:05:07.8145800Z Error in git ls-tree: exit status 128 fatal: not a tree object
2018-04-04T09:05:07.8152530Z 
2018-04-04T09:05:07.8160640Z 
```

There appears to be no pattern in the commits that fail. Retrying basically always works. It's just a weird intermittent issue. 

I'd love to give more information, but this is all we have unfortunately.

cc @dscho who I've spoken to about the VSTS side of things (they don't have any more info than we have).

**Comment by ttaylorr** [comment]
@dscho I think I gathered what's going on: can you try a build from the tip of [this][1] branch?

[1]: https://github.com/git-lfs/git-lfs/compare/git-include-sha1-in-ref-parse

**Comment by dscho** [comment]
@ttaylorr thanks! I will try to get that built and to @dalemyers for testing (I do not have any reproducer myself).

**Comment by ttaylorr** [comment]
> @ttaylorr, is this fix likely to go in anyway?

Yes, I'm going to merge this into master, and backport it to the next release in the v2.4.x series (as of writing, v2.4.1), since if at the very least this isn't fixing your bug, it will fix a bug. 😏

**Comment by dalemyers** [comment]
Looks like this was just merged. Thanks @ttaylorr! 

Is there a rough idea of when 2.4.1 might be out?

**Comment by ttaylorr** [comment]
> Is there a rough idea of when 2.4.1 might be out?

I'd like to release it next week, but I am waiting on a few topics to resolve before doing so.

**Comment by dalemyers** [comment]
Any updates on this? We have about 50% of our checkouts failing due to this so we are eager to get it in.

**Comment by dalemyers** [comment]
So I know that 2.4.1 was released 8 days ago, but we've had some delays in getting it onto our build machine images so we haven't been able to test this yet. However, even on 2.4.0 it seems that the number of these issues has dropped dramatically for no apparent reason. 

I'm going to be on vacation for a couple of weeks so I won't be able to report back until then.

**Comment by dalemyers** [comment]
So we are on 2.4.2 on our machines now and the issue still remains unfortunately. I've reached out to ask if any other teams are seeing the same issue (we share the same virtual machine pool), but so far it's just us. We are also trying to figure out a way we can test new builds to test solutions or just to debug. 

What can we do from here to figure out the issue?

**Comment by ttaylorr** [comment]
> What can we do from here to figure out the issue?

I'm a little puzzled that it's still happening. My best guess is that there are legitimately tree objects where there shouldn't be (or vice-versa). Perhaps you could run a strict `git fsck` before any Git LFS operation?

**Comment by dalemyers** [comment]
Would we not see anything like that locally by running it? For my local copy I'm seeing about 15 dangling blobs, 300 dangling commits and 15 dangling trees. Our repo is around 40,000 commits across all branches though, so I'm not too surprised by that. Should I be?

**Comment by ttaylorr** [comment]
Those do not sound surprising to me, so I think that they are OK. I'd be curious if you see anything out of the ordinary, though.

**Comment by dalemyers** [comment]
Nope. When we do a full clone we have around half a dozen missing files for lfs, but again, I don't think that's too surprising. What can we do to track down what's happening?

**Comment by ttaylorr** [comment]
@dalemyers Perhaps you can run it and report the output if/when the issue occurs again in the future? Without access to your repository, I am not sure of a way that I could be useful.

**Comment by dalemyers** [comment]
Ok, I'll see if we can get that to happen. It's gonna mean a change to VSTS so it could be quite tricky (or take a while). 

Once you guys are onboarded into Microsoft with a Microsoft email you'll be able to view our repo, so clearly that's what we need to push for! 😄

**Comment by dalemyers** [comment]
So we finally got to the bottom of this. It was user error basically. 

What was happening is that builds were triggered, but before they could always be cloned, the user force pushes, removing the commit so git can't find it and we see this error. That explains why it never happened when we watched it, only when people were doing their normal thing.

**Comment by ttaylorr** [comment]
@dalemyers great; I'm glad that things are all set. Please don't hesitate to open a new issue or ping me elsewhere in the future should anything arise. Thanks!
